apps/aecore/lib/aecore/contract/tx/contract_create_tx.ex

```diff
@@ -0,0 +1,286 @@
+
+    if length(senders) == 1 do
+      :ok
+    else
+      {:error, "#{__MODULE__}: Invalid senders number"}
+    end
+  end
+
+  @spec process_chainstate(
+          Chainstate.accounts(),
+          tx_type_state(),
+          non_neg_integer(),
+          t(),
+          DataTx.t(),
+          Transaction.context()
+        ) :: {:ok, {Chainstate.accounts(), tx_type_state()}}
+  def process_chainstate(
+        accounts,
+        contracts,
+        block_height,
+        %ContractCreateTx{
+          code: code,
+          vm_version: vm_version,
+          deposit: deposit,
+          amount: amount,
+          gas: gas,
+          gas_price: gas_price,
+          call_data: call_data
+        },
+        data_tx,
+        _context
+      ) do
+    owner = DataTx.main_sender(data_tx)
+    contract = Contract.new(owner, data_tx.nonce, vm_version, code, deposit)
+
+    updated_accounts_state =
+      accounts
+      |> AccountStateTree.update(owner, fn acc ->
+        Account.apply_transfer!(acc, block_height, amount * -1)
+      end)
+      |> AccountStateTree.update(contract.id.value, fn acc ->
+        Account.apply_transfer!(acc, block_height, amount)
+      end)
+
+    updated_contracts_state = ContractStateTree.insert_contract(contracts, contract)
+
+    call = Call.new(owner, data_tx.nonce, block_height, contract.id.value, gas_price)
+
+    call_definition = %{
+      caller: call.caller_address,
+      contract: contract.id,
+      gas: gas,
+      gas_price: gas_price,
+      call_data: call_data,
+      amount: amount,
+      call_stack: [],
+      code: contract.code,
+      call: call,
+      height: block_height
+    }
+
+    chain_state = %{
+      Chain.chain_state()
+      | contracts: updated_contracts_state,
+        accounts: updated_accounts_state
+    }
+
+    {call_result, updated_state} =
+      Dispatch.run(Constants.aevm_solidity_01(), call_definition, chain_state)
+
+    final_state =
+      case call_result.return_type do
+        return_type when return_type in [:ok, :revert] ->
+          gas_cost = call_result.gas_used * gas_price
+
+          accounts_after_gas_spent =
+            AccountStateTree.update(updated_accounts_state, owner, fn acc ->
+              Account.apply_transfer!(acc, block_height, (gas_cost + deposit) * -1)
+            end)
+
+          updated_contract = %{contract | code: call_result.return_value}
+
+          chain_state_with_call = %{
+            updated_state
+            | calls: CallStateTree.insert_call(updated_state.calls, call),
+              accounts: accounts_after_gas_spent,
+              contracts:
+                ContractStateTree.enter_contract(updated_state.contracts, updated_contract)
+          }
+
+        _error ->
+          gas_cost = call_result.gas_used * gas_price
+
+          accounts_after_gas_spent =
+            AccountStateTree.update(updated_accounts_state, owner, fn acc ->
+              Account.apply_transfer!(acc, block_height, gas_cost * -1)
```

Comment:
The `amount` transfer should also probably be reversed and not applied at all.

Reply:
I'm also confused about `amount` and `deposit`. Could you explain what are those and the difference between them?

Reply:
The `amount` is being sent to the contract account. The `deposit` is being held by the contract until it is deactivated.